### DISCUSSION TITLE
WAMR enable file system access

### DISCUSSION THREAD

**Discussion by m1cha1s** [bytecodealliance/wasm-micro-runtime#1473] [Q&A] [answered]
Has anybody attempted to allow WASI file access when embedding WAMR in C/C++ code? Because when i try to run a simple Rust program that opens a file and reads its contents compiled with: `cargo build --target wasm32-wasi`. Error message when trying to run it: `thread 'main' panicked at 'couldn't open test.txt: failed to find a pre-opened file descriptor through which "test.txt" could be opened', src/main.rs:10:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace`

Source files:
- [test.cpp](https://github.com/bytecodealliance/wasm-micro-runtime/files/9541405/test.txt)
- [main.rs](https://github.com/bytecodealliance/wasm-micro-runtime/files/9541406/main.txt)

**Comment by loganek** [comment, marked as answer]
@m1cha1s unfortunately your `test.cpp` only contains `Hello world` text.
Answering your question: your WASM code runs in sandboxing environment, therefore any I/O access must be explicitly granted. To enable access to specific directories, you can use the [`wasm_runtime_set_wasi_args`](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_export.h#L369) (second argument in that function is a list of directories that should be allowed, and the third one is the size of the list).

**Reply by m1cha1s** [reply]
@loganek thank you for the answer. I didn't find that function. It would be nice if there was a file access example, or the [wasm_runtime_set_wasi_args](https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/wasm_export.h#L369) description had some file access part or something.

**Comment by wenyongh** [comment]
@m1cha1s WASI is built using [capability-based security](https://github.com/WebAssembly/WASI/blob/main/README.md#capability-based-security) principles, to access the file under the folder, normally we should grant the acess privileges for the folder. For the `iwasm` built under product-mini/platforms/linux/build, we can use `iwasm --dir=<dir1> --dir=<dir2> ... <wasm file>`. For the usage of `wasm_runtime_set_wasi_args`, please refer to:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/product-mini/platforms/posix/main.c#L554-L555
Thanks.
